### ISSUE TITLE
OpenVPN GUI "Connection to management interface failed" error for domain user, with no local admin privileges

### ISSUE THREAD

**Issue by mason-ftl** [OpenVPN/openvpn-gui#332]
Client = OpenVPN GUI 11.14.0.0/2.4.8 (config and installer generated by pfSense client export plugin)
Server = OpenVPN on up-to-date pfSense
OS = Windows 10 Pro, feature pack 1803
Laptop = Lenovo ThinkPad E550

- User is a domain user with no local admin privileges
- OpenVPN GUI launches on Windows startup
- Config file and TLS key are located at C:\Program Files\OpenVPN\config
- OpenVPN Interactive Service is set to Automatic and is running
- OpenVPNService and OpenVPN Legacy Service are disabled

**Issue Description**
Sometimes, when double clicking on the OpenVPN GUI, the error message "Connecting to management interface failed" is displayed and no VPN connection is established.  However, sometimes the VPN connects normally.

**Steps to Consistently Reproduce the Issue**
- Reboot the PC
- Login as a regular domain user
- Double click on the OpenVPN GUI icon
- When prompted, enter username and password
- Wait for VPN to fully establish
- Disconnect from the VPN
- Wait until icon is no longer green
- Continue connecting and disconnecting until the error message is displayed, then click OK to close the error window.  Sometimes it only takes once, sometimes it takes 2 or 3 tries.
- Launch Task Manager as a local admin user
- Kill the OpenVPN Daemon process
- Wait at least 1 minute.  The issue does seem to be timing related, so it is important to wait a full 60 seconds.  I have tried immediately reconnecting (after killing the daemon) and often it will work.  I tried longer and longer delays until I found that it would never connect after 60 seconds.
- Double click on the OpenVPN GUI icon and wait for the error to pop up
- Again, if I kill the openvpn.exe process and immediately try to reconnect, it works.  But if I wait 60 seconds or more, then I again get the error message

**Steps Taken to Troubleshoot**
- Uninstalled client, deleted config, reinstalled client
- Confirmed, [as per this wiki entry](https://community.openvpn.net/openvpn/wiki/OpenVPN-GUI-New#gui-moreinfo), that the correct services were running
- Set verb 4 in config and used diff to compare a successful connection to an unsuccessful one. 
 They are identical until the "Enter Management Password:" entry in the logs, after which the successful connection continues, but there is nothing further from the failed connection.
- Suspected some sort of timeout threshold had been reached, so continued to insert longer and longer delays, following the killing of openvpn.exe, until I found the magic number of 1 minute.
- Tried logging in with a local admin user and following the above steps to reproduce the issue.  It is not possible to reproduce the issue when logged in as a **local admin** user.

**Client Config**
dev tun
persist-tun
persist-key
cipher AES-128-GCM
ncp-ciphers AES-128-GCM:AES-128-CBC
auth SHA512
tls-client
client
resolv-retry 2
remote vpn.example.com 1194 udp
setenv opt block-outside-dns
lport 0
verify-x509-name "hostname" name
auth-user-pass
pkcs12 fw1-udp-1194-hostname.p12
tls-auth fw1-udp-1194-hostname-tls.key 1
remote-cert-tls server
passtos
verb 4

**Comment by selvanair** [comment]
Sounds somewhat similar to the issue reported here:
https://community.openvpn.net/openvpn/ticket/1051

If the GUI fails to connect to management interface within 15 seconds after spawning oepnvpn.exe it times out. For some reason the core process may be taking a longer time to come up causing this behaviour. See the above link for more details.

The delay may be because user credentials are checked by the service before starting openvpn.exe and that may be taking long for domain accounts when the DC is unreachable or slow to respond. But the 60 second wait before restart causing issues sounds bizzare and  could be related to some Windows cache not controlled by OpenVPN.

Anyway, please try the exe in the proposed fix here:
https://github.com/selvanair/openvpn-gui/releases/tag/mgmt-timeout
This will to repeatedly try to connect to the management i/f instead of aborting, and may  fix the problem.

**Comment by mason-ftl** [comment]
Hello @selvanair,

Thank you so much for your quick response and for a solution that works!

While the fix does work (and I'm very thankful for it!), I think it may be masking an underlying problem.  I say this because the behaviour of a domain user with local admin privileges, is different than a domain user that has no local admin privileges.  When I'm logged into the laptop as a domain user with local admin privileges, I cannot reproduce the problem.  Yet, when logged in as a domain user with no local admin privileges, the issue is consistently reproducible.  In both cases, I frequently experience delays during the VPN connection process, which are much longer than if I'm connecting using a non-domain local admin user.

**Comment by selvanair** [comment]
When you have local admin privilege the interactive service considers you as authorized to use any connection profile, and no connection to DC is involved. Otherwise it has to check your group membership which can involve a connection to the DC with fallback to cached data. 

So the underlying problem is that either your DC is not reachable or is slow to respond. The patch takes care of such cases by not timing out in 15 seconds.

**Comment by mason-ftl** [comment]
The DC is unreachable because this laptop is a roaming device that is only able to connect to the DC once the VPN is established, so this will always be the case for this device.

If there is a way to decrease the DC connection timeout period or have the VPN client recognize that a domain user is attempting to connect, but there is no DC available?

**Comment by selvanair** [comment]
There could be some setting in the OS on how long to wait for the DC before falling back to the cached result. Even if you manage to control that,  the patch is required as its not correct, not required to assume that openvpn.exe will come up in 15 seconds.

But from a patch to a PR takes a lot of due-diligence and even for submitted PRs we are short of people to review and ACK  them, so no idea when this will get submitted, reviewed and merged.

**Comment by selvanair** [comment]
By the way, its possible to eliminate the group membership check (and connection attempt to DC) if the config is in the global directory (not in user profile). I'll submit a patch to openvpn core for this.

**Comment by mason-ftl** [comment]
I assume by "global directory" you mean C:\Program Files\OpenVPN\config.  That's where my configs are located.

**Comment by selvanair** [comment]
@mason-ftl Yes, no further check is necessary in that case (but still done in the current code) and there will be virtually no delay in spawning openvpn.exe by the service or for the management i/f to come up. Delays in establishing connection with the server is irrelevant here.

**Comment by mason-ftl** [comment]
Excellent!  That seems like a great interim solution to the problem, while you're waiting for your other patch to make it's way into core.

Thanks @selvanair!

**Comment by selvanair** [comment]
Keep this open until the issue is resolved.
I've submitted a patch for openvpn service (https://patchwork.openvpn.net/patch/969/)

**Comment by selvanair** [comment]
The patch for interactive service has been merged and will be available in the 2.5 release.

To test whether it fixes the issue, try the snapshot build. The latest at this moment is https://build.openvpn.net/downloads/snapshots/openvpn-install-2.5_git-I900-master-20200213201015-aa6affe6.exe

**Comment by cron2** [comment]
Hi,

On Mon, Feb 17, 2020 at 03:22:21PM -0800, Selva Nair wrote:
> The patch for interactive service has been merged and will be available in the 2.5 release.

Shall we pull this up to the next 2.4?   If I am remembering correctly,
this is a "standalone" thing, with no dependencies on other changes.

gert
-- 
"If was one thing all people took for granted, was conviction that if you 
 feed honest figures into a computer, honest figures come out. Never doubted 
 it myself till I met a computer with a sense of humor."
                             Robert A. Heinlein, The Moon is a Harsh Mistress

Gert Doering - Munich, Germany                             gert@greenie.muc.de

**Comment by selvanair** [comment]
On Tue, Feb 18, 2020 at 7:03 AM Gert Doering <notifications@github.com>
wrote:

> Hi,
>
> On Mon, Feb 17, 2020 at 03:22:21PM -0800, Selva Nair wrote:
> > The patch for interactive service has been merged and will be available
> in the 2.5 release.
>
> Shall we pull this up to the next 2.4? If I am remembering correctly,
> this is a "standalone" thing, with no dependencies on other changes.
>

Yes, this would be useful in 2.4 and there are no real functionality
changes.

Unfortunately it doesn't cherry-pick cleanly because of past swprintf ->
openvpn_swprintf
changes. Resolution is easy, but I can send a 2.4 specific patch if
preferred.

Selva

**Comment by cron2** [comment]
Hi,

On Tue, Feb 18, 2020 at 07:46:08AM -0800, Selva Nair wrote:
> Resolution is easy, but I can send a 2.4 specific patch if preferred.

As I'm lazy *and* very busy these weeks, this would be nice :-)

gert
-- 
"If was one thing all people took for granted, was conviction that if you 
 feed honest figures into a computer, honest figures come out. Never doubted 
 it myself till I met a computer with a sense of humor."
                             Robert A. Heinlein, The Moon is a Harsh Mistress

Gert Doering - Munich, Germany                             gert@greenie.muc.de

**Comment by CryptoSiD** [comment]
I'm still having this issue with the latest version of OpenVPN (2.5.2)

**Comment by selvanair** [comment]
> I'm still having this issue with the latest version of OpenVPN (2.5.2)

This issue means what? There are number of things discussed here related to location of config, user local admin or not, DC reachable or not, time out not long enough etc.

**Comment by CryptoSiD** [comment]
> > I'm still having this issue with the latest version of OpenVPN (2.5.2)
> 
> This issue means what? There are number of things discussed here related to location of config, user local admin or not, DC reachable or not, timeout not long enough etc.

Hi, sorry I should have been more precise.

I had the error "Connection to management interface failed" when trying to connect to any VPN with OpenVPN GUI.

However, after a reboot, it started working properly.

You can keep this closed :)

**Comment by selvanair** [comment]
> However, after a reboot, it started working properly.

There is nothin' that a little "reboot" can't fix :)

**Comment by gotexis** [comment]
This shall be reopened

**Comment by PrakashLakhara** [comment]
However, there is a temp fix:
```
1. Kill OpenVPN processes from the task manager.
2. Disable and Enable TAP network adapter.
3. Then try to connect.
```

**Comment by Yonisen** [comment]
> Was getting the error. UnInstalled OpenVPN. Went through the Registry and deleted anything pertaining to OpenVPN and then reinstalled.

You guy real genius, i did this (just find by "openvpn" in Registry) and its work! This is the only thing that helped me.

**Comment by scscgit** [comment]
> > However, after a reboot, it started working properly.
> 
> There is nothin' that a little "reboot" can't fix :)

  Is there any plan to address the root cause though? How can we troubleshoot this in real time?

**Comment by cron2** [comment]
> Is there any plan to address the root cause though? How can we troubleshoot this in real time?

The root cause for what exactly?  There are a number of different problem people have dropped in this ticket...

The issue in the title, related to DC lookups not working for local users because the VPN is not up yet, has been solved over 5 years ago.

As this ticket is a mixture of one real (and fixed) problem, and a number of other non-useful comments dropped in, I think the best approach would be to open a new issue, and describe exactly what problem you are seeing, on what windows version, and with which OpenVPN version.  Feel free to refer here.

**Comment by scscgit** [comment]
> > Is there any plan to address the root cause though? How can we troubleshoot this in real time?
> 
> The root cause for what exactly? There are a number of different problem people have dropped in this ticket...

I just generally mean the error of infinitely repeating error "Connection to management interface failed". The fact that it repeats means the solution of automatic retrying (mentioned here) isn't sufficient.

I've just now had some weird problem with an antivirus, which generally caused some network issues on my machine, and this is just yet another problem that appeared afterward. I can't access the VPN, however many times I restart the service. I'm pretty sure the problem will get resolved by a restart, so I don't have any repro steps to provide to be able to help developers debug this on their own, meaning it's unproductive to create a new ticket for the sake of something I can't even define properly, but it's a fact that OpenVPN can enter an unstable state where this error is displayed. My OpenVPN installation isn't older than 3 years. This is the only thread where any discussion exists AFAIK, so I have no other place to leave this note. I don't really expect any fix, so I'm just soliciting any troubleshooting/debugging tips in case this problem reappears years later for other people as well. (I'm categorically against considering a computer restart to be a solution.) Thanks anyway :)

**Comment by cron2** [comment]
> (I'm categorically against considering a computer restart to be a solution.)

Well, you're using *Windows*... plus some 3rd party antivirus software...

But anyway, thanks for confirming that you are indeed referring to the original topic of this issue.  Not sure how to troubleshoot this - windows event log might have something, or the gui log.

**Comment by scscgit** [comment]
The `log/edge-`...`.log` logs nothing new after the 3 startup initialization lines. GUI just keeps retrying with the same error. And the Application view of the Windows Logs only logs an initial
```
<EventData>
  <Data>openvpnserv:</Data> 
  <Data>The Authorizing user '<username>' by virtue of membership in group 'Administrators'</Data> 
</EventData>
```
I doubt this is as easy as reading through logs :) In any case, it's not like there is a better/safer work environment than _Windows + 3rd party antivirus_; sadly, that's the state of the world we live in. And given enough anger, one would be willing to even learn to decompile the networking driver stack for the sake of brute-forcing any solution, but well, I guess the necessary resources for Windows aren't so easily found :) So next time this happens again, I won't find any solution either.

Edited to reduce the number of messages:

With `verb 4` in the config, these are the last log lines. I'm not sure if any specific certificate parameters or versioning details are relevant here; I'm not on the latest version yet, so if this behavior has changed recently, then it's fine as I'll update soon enough:

OpenVPN 2.5.(...) Windows-MSVC
Windows version 10.0 (Windows 10 or greater) 64bit
library versions: (...)
MANAGEMENT: TCP Socket listening on [AF_INET]127.0.0.1:25340
Need hold release from management interface, waiting...

I don't have any duplicate processes, I've restarted the service too.

**Comment by selvanair** [comment]
> And given enough anger, one would be willing to even learn to decompile the networking driver stack for the sake of brute-forcing any solution, but well, I guess the necessary resources for Windows aren't so easily found :) So next time this happens again, I won't find any solution either.

Your are overthinking it. The ~error~ event log info message shows user is already authorized and so this is unelated to the issue discussed in this thread (scroll up and see that this was about unreachable DC causing user group lookup failure).

Increase verbosity to 4 (verb 4 in config file) and post the logs here. And tell us the version of OpenVPN-GUI you are using. Also check whether multiple openvpn.exe instances are running when this error occurs. There could be stale executable(s) left behind by an unsuccessful connection which makes troubleshooting harder.
